// sanity/lib/_chunks/devAction-c2147bc0.js
'use strict';

var path = require('path');
var chalk = require('chalk');
var vite = require('vite');
var runtime = require('./runtime-87e16971.js');
var timing = require('./timing-612e5342.js');
var servers = require('./servers-214c2a6f.js');
function _interopDefaultCompat(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    default: e
  };
}
var path__default = /*#__PURE__*/_interopDefaultCompat(path);
var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);
async function startDevServer(options) {
  const {
    cwd,
    httpPort,
    httpHost,
    basePath: base,
    reactStrictMode,
    vite: extendViteConfig
  } = options;
  const startTime = Date.now();
  runtime.debug("Writing Sanity runtime files");
  await runtime.writeSanityRuntime({
    cwd,
    reactStrictMode,
    watch: true
  });
  runtime.debug("Resolving vite config");
  let viteConfig = await runtime.getViteConfig({
    basePath: base || "/",
    mode: "development",
    server: {
      port: httpPort,
      host: httpHost
    },
    cwd
  });
  if (extendViteConfig) {
    runtime.debug("Extending vite config using user-specified function");
    viteConfig = extendViteConfig(viteConfig);
  }
  runtime.debug("Creating vite server");
  const server = await vite.createServer(viteConfig);
  const info = server.config.logger.info;
  runtime.debug("Listening on specified port");
  await server.listen();
  const startupDuration = Date.now() - startTime;
  const url = "http://".concat(httpHost || "localhost", ":").concat(httpPort || "3333");
  info("Sanity Studio using ".concat(chalk__default.default.cyan("vite@".concat(require("vite/package.json").version)), " ready in ").concat(chalk__default.default.cyan("".concat(Math.ceil(startupDuration), "ms")), " and running at ").concat(chalk__default.default.cyan(url)));
  return {
    close: () => server.close()
  };
}
async function startSanityDevServer(args, context) {
  const timers = timing.getTimer();
  const flags = args.extOptions;
  const {
    output,
    workDir,
    cliConfig
  } = context;
  timers.start("checkStudioDependencyVersions");
  timing.checkStudioDependencyVersions(workDir);
  timers.end("checkStudioDependencyVersions");
  if ((await timing.checkRequiredDependencies(context)).didInstall) {
    return;
  }
  const configSpinner = output.spinner("Checking configuration files...");
  const config = getDevServerConfig({
    flags,
    workDir,
    cliConfig
  });
  configSpinner.succeed();
  try {
    await startDevServer(config);
  } catch (err) {
    servers.gracefulServerDeath("dev", config.httpHost, config.httpPort, err);
  }
}
function getDevServerConfig(_ref) {
  let {
    flags,
    workDir,
    cliConfig
  } = _ref;
  const baseConfig = servers.getSharedServerConfig({
    flags,
    workDir,
    cliConfig
  });
  const env = process.env;
  const reactStrictMode = env.SANITY_STUDIO_REACT_STRICT_MODE ? env.SANITY_STUDIO_REACT_STRICT_MODE === "true" : Boolean(cliConfig == null ? void 0 : cliConfig.reactStrictMode);
  return {
    ...baseConfig,
    staticPath: path__default.default.join(workDir, "static"),
    reactStrictMode
  };
}
exports.default = startSanityDevServer;
