// sanity/lib/_chunks/index-1e1252eb.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { useState, useEffect, useCallback, useMemo, memo, useRef } from 'react';
import { Flex, Container, Stack, Heading, Text, Box, Button, Spinner, VirtualList, Card } from '@sanity/ui';
import shallowEquals from 'shallow-equals';
import { SyncIcon, ArrowLeftIcon } from '@sanity/icons';
import 'sanity/router';
import { usePaneLayout, usePane, PaneContent, PaneHeader, BackLink, PaneHeaderActions, Pane, _DEBUG } from './desk-eb7b3548.js';
import { Delay, useDeskTool, useDeskToolSetting } from './index-66ec8894.js';
import * as PathUtils from '@sanity/util/paths';
import { isKeySegment, isIndexSegment, isReferenceSchemaType } from '@sanity/types';
import { getPublishedId, collate, useSchema, listenQuery, useClient, DEFAULT_STUDIO_CLIENT_OPTIONS, useSource, useUnique, SourceProvider } from 'sanity';
import { PaneItem } from './PaneItem-59fdbcf0.js';
import 'lodash/omit.js';
import 'lodash/isNumber.js';
import 'lodash/isString.js';
import 'react-rx';
import { Subject, merge, of, concat, fromEvent } from 'rxjs';
import { map, distinctUntilChanged, publishReplay, refCount, switchMap, share, delay, takeUntil, startWith, catchError, take, mergeMapTo, scan, filter } from 'rxjs/operators';
import deepCompare from 'react-fast-compare';
const PARTIAL_PAGE_LIMIT = 100;
const FULL_LIST_LIMIT = 2e3;
const DEFAULT_ORDERING = {
  by: [{
    field: "_updatedAt",
    direction: "desc"
  }]
};
const EMPTY_RECORD = {};
function getDocumentKey(value, index) {
  return value._id ? getPublishedId(value._id) : "item-".concat(index);
}
function removePublishedWithDrafts(documents) {
  return collate(documents).map(entry => {
    const doc = entry.draft || entry.published;
    return {
      ...doc,
      hasPublished: !!entry.published,
      hasDraft: !!entry.draft
    };
  });
}
const RE_TYPE_NAME_IN_FILTER = /\b_type\s*==\s*(['"].*?['"]|\$.*?(?:\s|$))|\B(['"].*?['"]|\$.*?(?:\s|$))\s*==\s*_type\b/;
function getTypeNameFromSingleTypeFilter(filter) {
  let params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const matches = filter.match(RE_TYPE_NAME_IN_FILTER);
  if (!matches) {
    return null;
  }
  const match = (matches[1] || matches[2]).trim().replace(/^["']|["']$/g, "");
  if (match[0] === "$") {
    const k = match.slice(1);
    const v = params[k];
    return typeof v === "string" ? v : null;
  }
  return match;
}
function isSimpleTypeFilter(filter) {
  return /^_type\s*==\s*['"$]\w+['"]?\s*$/.test(filter.trim());
}
function toOrderClause(orderBy) {
  return orderBy.map(ordering => [wrapFieldWithFn(ordering), (ordering.direction || "").toLowerCase()].map(str => str.trim()).filter(Boolean).join(" ")).join(",");
}
function wrapFieldWithFn(ordering) {
  return ordering.mapWith ? "".concat(ordering.mapWith, "(").concat(ordering.field, ")") : ordering.field;
}
function applyOrderingFunctions(order, schemaType) {
  const orderBy = order.by.map(by => {
    if (by.mapWith) {
      return by;
    }
    const fieldType = tryResolveSchemaTypeForPath(schemaType, by.field);
    if (!fieldType) {
      return by;
    }
    if (fieldExtendsType(fieldType, "datetime")) {
      return {
        ...by,
        mapWith: "dateTime"
      };
    }
    if (fieldType.jsonType === "string") {
      return {
        ...by,
        mapWith: "lower"
      };
    }
    return by;
  });
  return orderBy.every((item, index) => item === order.by[index]) ? order : {
    ...order,
    by: orderBy
  };
}
function tryResolveSchemaTypeForPath(baseType, path) {
  const pathSegments = PathUtils.fromString(path);
  let current = baseType;
  for (const segment of pathSegments) {
    if (!current) {
      return void 0;
    }
    if (typeof segment === "string") {
      current = getFieldTypeByName(current, segment);
      continue;
    }
    const isArrayAccessor = isKeySegment(segment) || isIndexSegment(segment);
    if (!isArrayAccessor || current.jsonType !== "array") {
      return void 0;
    }
    const [memberType, otherType] = current.of || [];
    if (otherType || !memberType) {
      return void 0;
    }
    if (!isReferenceSchemaType(memberType)) {
      current = memberType;
      continue;
    }
    const [refType, otherRefType] = memberType.to || [];
    if (otherRefType || !refType) {
      return void 0;
    }
    current = refType;
  }
  return current;
}
function getFieldTypeByName(type, fieldName) {
  if (!("fields" in type)) {
    return void 0;
  }
  const fieldType = type.fields.find(field => field.name === fieldName);
  return fieldType ? fieldType.type : void 0;
}
function fieldExtendsType(field, ofType) {
  let current = field.type;
  while (current) {
    if (current.name === ofType) {
      return true;
    }
    if (!current.type && current.jsonType === ofType) {
      return true;
    }
    current = current.type;
  }
  return false;
}
function DocumentListPaneContent(props) {
  const {
    childItemId,
    error,
    filterIsSimpleTypeContraint,
    fullList,
    isActive,
    isLoading,
    items,
    layout,
    onListChange,
    onRetry,
    showIcons
  } = props;
  const schema = useSchema();
  const {
    collapsed: layoutCollapsed
  } = usePaneLayout();
  const {
    collapsed,
    index
  } = usePane();
  const [shouldRender, setShouldRender] = useState(false);
  useEffect(() => {
    if (collapsed) return void 0;
    const timer = setTimeout(() => {
      setShouldRender(true);
    }, 0);
    return () => {
      clearTimeout(timer);
    };
  }, [collapsed]);
  const renderItem = useCallback(item => {
    const publishedId = getPublishedId(item._id);
    const isSelected = childItemId === publishedId;
    const pressed = !isActive && isSelected;
    const selected = isActive && isSelected;
    return /* @__PURE__ */jsx(PaneItem, {
      icon: showIcons === false ? false : void 0,
      id: publishedId,
      pressed,
      selected,
      layout,
      schemaType: schema.get(item._type),
      value: item
    });
  }, [childItemId, isActive, layout, schema, showIcons]);
  const content = useMemo(() => {
    if (!shouldRender) {
      return null;
    }
    if (error) {
      return /* @__PURE__ */jsx(Flex, {
        align: "center",
        direction: "column",
        height: "fill",
        justify: "center",
        children: /* @__PURE__ */jsx(Container, {
          width: 1,
          children: /* @__PURE__ */jsxs(Stack, {
            paddingX: 4,
            paddingY: 5,
            space: 4,
            children: [/* @__PURE__ */jsx(Heading, {
              as: "h3",
              children: "Could not fetch list items"
            }), /* @__PURE__ */jsxs(Text, {
              as: "p",
              children: ["Error: ", /* @__PURE__ */jsx("code", {
                children: error.message
              })]
            }), onRetry && /* @__PURE__ */jsx(Box, {
              children: /* @__PURE__ */jsx(Button, {
                icon: SyncIcon,
                onClick: onRetry,
                text: "Retry",
                tone: "primary"
              })
            })]
          })
        })
      });
    }
    if (items === null) {
      return /* @__PURE__ */jsx(Flex, {
        align: "center",
        direction: "column",
        height: "fill",
        justify: "center",
        children: /* @__PURE__ */jsx(Delay, {
          ms: 300,
          children: /* @__PURE__ */jsxs(Fragment, {
            children: [/* @__PURE__ */jsx(Spinner, {
              muted: true
            }), /* @__PURE__ */jsx(Box, {
              marginTop: 3,
              children: /* @__PURE__ */jsx(Text, {
                align: "center",
                muted: true,
                size: 1,
                children: "Loading documents\u2026"
              })
            })]
          })
        })
      });
    }
    if (!isLoading && items.length === 0) {
      return /* @__PURE__ */jsx(Flex, {
        align: "center",
        direction: "column",
        height: "fill",
        justify: "center",
        children: /* @__PURE__ */jsx(Container, {
          width: 1,
          children: /* @__PURE__ */jsx(Box, {
            paddingX: 4,
            paddingY: 5,
            children: /* @__PURE__ */jsx(Text, {
              align: "center",
              muted: true,
              size: 2,
              children: filterIsSimpleTypeContraint ? "No documents of this type" : "No matching documents"
            })
          })
        })
      });
    }
    const hasMoreItems = fullList && items.length === FULL_LIST_LIMIT;
    return /* @__PURE__ */jsxs(Box, {
      padding: 2,
      children: [items.length > 0 && /* @__PURE__ */jsx(VirtualList, {
        gap: 1,
        getItemKey: getDocumentKey,
        items,
        renderItem,
        onChange: onListChange
      }, "".concat(index, "-").concat(collapsed)), isLoading && /* @__PURE__ */jsx(Card, {
        borderTop: true,
        marginTop: 1,
        paddingX: 3,
        paddingY: 4,
        children: /* @__PURE__ */jsx(Text, {
          align: "center",
          muted: true,
          size: 1,
          children: "Loading\u2026"
        })
      }), hasMoreItems && /* @__PURE__ */jsx(Card, {
        marginTop: 1,
        paddingX: 3,
        paddingY: 4,
        radius: 2,
        tone: "transparent",
        children: /* @__PURE__ */jsxs(Text, {
          align: "center",
          muted: true,
          size: 1,
          children: ["Displaying a maximum of ", FULL_LIST_LIMIT, " documents"]
        })
      })]
    });
  }, [error, filterIsSimpleTypeContraint, fullList, onListChange, isLoading, items, onRetry, renderItem, shouldRender, collapsed, index]);
  return /* @__PURE__ */jsx(PaneContent, {
    overflow: layoutCollapsed ? void 0 : "auto",
    children: content
  });
}
const DocumentListPaneHeader = memo(_ref => {
  let {
    index,
    initialValueTemplates = [],
    menuItems = [],
    menuItemGroups = [],
    setLayout,
    setSortOrder,
    title
  } = _ref;
  const {
    features
  } = useDeskTool();
  const actionHandlers = useMemo(() => {
    return {
      setLayout: _ref2 => {
        let {
          layout: value
        } = _ref2;
        setLayout(value);
      },
      setSortOrder: sort => {
        setSortOrder(sort);
      }
    };
  }, [setLayout, setSortOrder]);
  return /* @__PURE__ */jsx(PaneHeader, {
    backButton: features.backButton && index > 0 && /* @__PURE__ */jsx(Button, {
      as: BackLink,
      "data-as": "a",
      icon: ArrowLeftIcon,
      mode: "bleed"
    }),
    title,
    actions: /* @__PURE__ */jsx(PaneHeaderActions, {
      initialValueTemplateItems: initialValueTemplates,
      actionHandlers,
      menuItemGroups,
      menuItems
    })
  });
});
DocumentListPaneHeader.displayName = "DocumentListPaneHeader";
const INITIAL_CHILD_PROPS = {
  result: null,
  error: false
};
const createResultChildProps = documents => ({
  result: {
    documents
  },
  loading: false,
  error: false
});
const createErrorChildProps = error => ({
  result: null,
  loading: false,
  error
});
const getQueryResults = function (receivedProps$) {
  let options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const onRetry$ = new Subject();
  const onRetry = onRetry$.next.bind(onRetry$);
  const queryProps$ = receivedProps$.pipe(map(props => ({
    client: props.client,
    query: props.query,
    params: props.params
  })), distinctUntilChanged(deepCompare), publishReplay(1), refCount());
  const queryResults$ = queryProps$.pipe(switchMap(queryProps => {
    const query$ = listenQuery(queryProps.client, queryProps.query, queryProps.params, options).pipe(map(createResultChildProps), share());
    return merge(of({
      loading: true
    }).pipe(delay(400), takeUntil(query$)), query$);
  }));
  return queryResults$.pipe(startWith(INITIAL_CHILD_PROPS), catchError((err, caught$) => concat(of(createErrorChildProps(err)), merge(fromEvent(window, "online"), onRetry$).pipe(take(1), mergeMapTo(caught$)))), scan((prev, next) => ({
    ...prev,
    ...next,
    onRetry
  })));
};
function useDocumentList(opts) {
  var _a;
  const {
    apiVersion,
    filter: filter$1,
    params,
    sortOrder
  } = opts;
  const client = useClient(DEFAULT_STUDIO_CLIENT_OPTIONS);
  const [fullList, setFullList] = useState(false);
  const fullListRef = useRef(fullList);
  const [result, setResult] = useState(null);
  const error = (result == null ? void 0 : result.error) || null;
  const isLoading = (result == null ? void 0 : result.loading) || result === null;
  const onRetry = result == null ? void 0 : result.onRetry;
  const documents = (_a = result == null ? void 0 : result.result) == null ? void 0 : _a.documents;
  const items = useMemo(() => documents ? removePublishedWithDrafts(documents) : null, [documents]);
  const query = useMemo(() => {
    const extendedProjection = sortOrder == null ? void 0 : sortOrder.extendedProjection;
    const projectionFields = ["_id", "_type"];
    const finalProjection = projectionFields.join(",");
    const sortBy = (sortOrder == null ? void 0 : sortOrder.by) || [];
    const limit = fullList ? FULL_LIST_LIMIT : PARTIAL_PAGE_LIMIT;
    const sort = sortBy.length > 0 ? sortBy : DEFAULT_ORDERING.by;
    const order = toOrderClause(sort);
    if (extendedProjection) {
      const firstProjection = projectionFields.concat(extendedProjection).join(",");
      return ["*[".concat(filter$1, "] {").concat(firstProjection, "}"), "order(".concat(order, ") [0...").concat(limit, "]"), "{".concat(finalProjection, "}")].join("|");
    }
    return "*[".concat(filter$1, "]|order(").concat(order, ")[0...").concat(limit, "]{").concat(finalProjection, "}");
  }, [filter$1, fullList, sortOrder]);
  const handleListChange = useCallback(_ref3 => {
    let {
      toIndex
    } = _ref3;
    if (isLoading || fullListRef.current) {
      return;
    }
    if (toIndex >= PARTIAL_PAGE_LIMIT / 2) {
      setFullList(true);
      fullListRef.current = true;
    }
  }, [isLoading]);
  useEffect(() => {
    const filterFn = fullList ? queryResult => Boolean(queryResult.result) : () => true;
    setResult(r => r ? {
      ...r,
      loading: true
    } : null);
    const queryResults$ = getQueryResults(of({
      client,
      query,
      params
    }), {
      apiVersion,
      tag: "desk.document-list"
    }).pipe(filter(filterFn));
    const sub = queryResults$.subscribe(setResult);
    return () => sub.unsubscribe();
  }, [apiVersion, client, fullList, query, params]);
  useEffect(() => {
    setResult(null);
    setFullList(false);
    fullListRef.current = false;
  }, [filter$1, params, sortOrder, apiVersion]);
  return {
    error,
    fullList,
    handleListChange,
    isLoading,
    items,
    onRetry
  };
}
const emptyArray = [];
function useShallowUnique(value) {
  const valueRef = useRef(value);
  if (!shallowEquals(valueRef.current, value)) {
    valueRef.current = value;
  }
  return valueRef.current;
}
const DocumentListPane = memo(function DocumentListPane2(props) {
  const {
    childItemId,
    index,
    isActive,
    isSelected,
    pane,
    paneKey
  } = props;
  const schema = useSchema();
  const {
    name: parentSourceName
  } = useSource();
  const {
    defaultLayout = "default",
    displayOptions,
    initialValueTemplates = emptyArray,
    menuItems,
    menuItemGroups,
    options,
    title
  } = pane;
  const {
    apiVersion,
    defaultOrdering = emptyArray,
    filter
  } = options;
  const params = useShallowUnique(options.params || EMPTY_RECORD);
  const sourceName = pane.source;
  const typeName = useMemo(() => getTypeNameFromSingleTypeFilter(filter, params), [filter, params]);
  const showIcons = (displayOptions == null ? void 0 : displayOptions.showIcons) !== false;
  const [layout, setLayout] = useDeskToolSetting(typeName, "layout", defaultLayout);
  const defaultSortOrder = useMemo(() => {
    return (defaultOrdering == null ? void 0 : defaultOrdering.length) > 0 ? {
      by: defaultOrdering
    } : DEFAULT_ORDERING;
  }, [defaultOrdering]);
  const [sortOrderRaw, setSortOrder] = useDeskToolSetting(typeName, "sortOrder", defaultSortOrder);
  const sortWithOrderingFn = typeName && sortOrderRaw ? applyOrderingFunctions(sortOrderRaw, schema.get(typeName)) : sortOrderRaw;
  const sortOrder = useUnique(sortWithOrderingFn);
  const filterIsSimpleTypeContraint = isSimpleTypeFilter(filter);
  const {
    error,
    fullList,
    handleListChange,
    isLoading,
    items,
    onRetry
  } = useDocumentList({
    filter,
    params,
    sortOrder,
    apiVersion
  });
  return /* @__PURE__ */jsx(SourceProvider, {
    name: sourceName || parentSourceName,
    children: /* @__PURE__ */jsxs(Pane, {
      currentMaxWidth: 350,
      id: paneKey,
      maxWidth: 640,
      minWidth: 320,
      selected: isSelected,
      children: [_DEBUG, /* @__PURE__ */jsx(DocumentListPaneHeader, {
        index,
        initialValueTemplates,
        menuItems,
        menuItemGroups,
        setLayout,
        setSortOrder,
        title
      }), /* @__PURE__ */jsx(DocumentListPaneContent, {
        childItemId,
        error,
        filterIsSimpleTypeContraint,
        fullList,
        isActive,
        isLoading,
        items,
        layout,
        onListChange: handleListChange,
        onRetry,
        showIcons
      })]
    })
  });
});
export { DocumentListPane as default };
