// get-it/lib/middleware/keepAlive.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _http = _interopRequireDefault(require("http"));

var _https = _interopRequireDefault(require("https"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var isHttpsProto = /^https:/i;

var _default = function _default() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var ms = config.ms || 1000;
  var maxFree = config.maxFree || 256;
  var agentOptions = {
    keepAlive: true,
    keepAliveMsecs: ms,
    maxFreeSockets: maxFree
  };
  var httpAgent = new _http.default.Agent(agentOptions);
  var httpsAgent = new _https.default.Agent(agentOptions);
  var agents = {
    http: httpAgent,
    https: httpsAgent
  };
  return {
    finalizeOptions: function finalizeOptions(options) {
      if (options.agent) {
        return options;
      }

      var isHttps = isHttpsProto.test(options.href || options.protocol);
      var keepOpts = options.maxRedirects === 0 ? {
        agent: isHttps ? httpsAgent : httpAgent
      } : {
        agents: agents
      };
      return _extends({}, options, keepOpts);
    }
  };
};

exports.default = _default;
module.exports = exports.default;
