// get-it/lib-node/middleware/keepAlive.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _http = _interopRequireDefault(require("http"));

var _https = _interopRequireDefault(require("https"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const isHttpsProto = /^https:/i;

var _default = (config = {}) => {
  const ms = config.ms || 1000;
  const maxFree = config.maxFree || 256;
  const agentOptions = {
    keepAlive: true,
    keepAliveMsecs: ms,
    maxFreeSockets: maxFree
  };
  const httpAgent = new _http.default.Agent(agentOptions);
  const httpsAgent = new _https.default.Agent(agentOptions);
  const agents = {
    http: httpAgent,
    https: httpsAgent
  };
  return {
    finalizeOptions: options => {
      if (options.agent) {
        return options;
      }

      const isHttps = isHttpsProto.test(options.href || options.protocol);
      const keepOpts = options.maxRedirects === 0 ? {
        agent: isHttps ? httpsAgent : httpAgent
      } : {
        agents
      };
      return Object.assign({}, options, keepOpts);
    }
  };
};

exports.default = _default;
module.exports = exports.default;
